### PR TITLE
#1886 improve Dynamic Time Warping distance eval for nrt

### LINKED ISSUE #1886: Optimization of  DTW of nrt
Average time gain of 30 %

### PR DESCRIPTION
**Pull request by marcochiaramello**
Closes #1886

### CODE REVIEW

File: util/nrt_diff/nrtDiff.py [hunk 1]
```diff
@@ -1265,17 +1265,17 @@ def XMLCloseEnough (path_left, path_right):
 def DTWDistance(left, right) :
     n = len(left)
     m = len(right)
-    DTW = [[0 for j in range(m+1)] for i in range(n+1)]
-
-    for i in range(1, n+1):
-        for j in range (1, m+1):
-            DTW[i][j] = 999999
+    DTW = [[999999 for j in range(m+1)] for i in range(n+1)]
```
Comment (lecourtoisflo) [inline util/nrt_diff/nrtDiff.py:1268]: We don't put 0 in first line and first column, only the first case. is it normal ?

**Review by lecourtoisflo — APPROVED** [review]

File: util/nrt_diff/nrtDiff.py (hunk 1 above)
Comment (marcochiaramello) [reply util/nrt_diff/nrtDiff.py:1268]: yes, you are right, by chance it do not impacted a lot, it is now fixed

**Review by lecourtoisflo — APPROVED** [review]

File: util/nrt_diff/nrtDiff.py [hunk 2]
```diff
@@ -1265,17 +1265,20 @@ def XMLCloseEnough (path_left, path_right):
...
     m = len(right)
-    DTW = [[0 for j in range(m+1)] for i in range(n+1)]
-
-    for i in range(1, n+1):
-        for j in range (1, m+1):
-            DTW[i][j] = 999999
+    if n == m:
+        if left == right:
+            return 0.
+    DTW = [[0 if (j == 0 and i == 0) else 999999 for j in range(m+1)] for i in range(n+1)]
```
Comment (lecourtoisflo) [inline util/nrt_diff/nrtDiff.py:1271]: shouldn't it be j == 0 or i == 0 ? ( first line or first column)

File: util/nrt_diff/nrtDiff.py [hunk 3]
```diff
@@ -1265,17 +1265,20 @@ def XMLCloseEnough (path_left, path_right):
...
     DTW[0][0] = 0
```
Comment (lecourtoisflo) [inline util/nrt_diff/nrtDiff.py:1272]: Is this line still required ?

File: util/nrt_diff/nrtDiff.py [hunk 4]
```diff
@@ -1265,17 +1265,20 @@ def XMLCloseEnough (path_left, path_right):
 def DTWDistance(left, right) :
     n = len(left)
     m = len(right)
-    DTW = [[0 for j in range(m+1)] for i in range(n+1)]
-
-    for i in range(1, n+1):
-        for j in range (1, m+1):
-            DTW[i][j] = 999999
+    if n == m:
```
Comment (lecourtoisflo) [inline util/nrt_diff/nrtDiff.py:1268]: why check the size if you check equality just after ?

**Review by lecourtoisflo — CHANGES_REQUESTED** [review]

File: util/nrt_diff/nrtDiff.py (hunk 2 above)
Comment (marcochiaramello) [reply util/nrt_diff/nrtDiff.py:1271]: yes, sorry again :)

File: util/nrt_diff/nrtDiff.py (hunk 3 above)
Comment (marcochiaramello) [reply util/nrt_diff/nrtDiff.py:1272]: no

File: util/nrt_diff/nrtDiff.py (hunk 4 above)
Comment (marcochiaramello) [reply util/nrt_diff/nrtDiff.py:1268]: I do not know what python does if comparing two different sizes lists, but anyway, if sizes are different, we need to go throught DTW

Comment (lecourtoisflo) [reply util/nrt_diff/nrtDiff.py:1268]: comparing lists compare also the size in python

**Review by lecourtoisflo — CHANGES_REQUESTED** [review]

**Review by lecourtoisflo — APPROVED** [review]

**Review by cossartque — APPROVED** [review]

**Pull request merged**